lib/Runtime/Library/JavascriptArray.cpp

```diff
@@ -6676,367 +6676,326 @@ using namespace Js;
+                            list[--k] = leftElement;
+                            if (i > left)
+                            {
+                                leftElement = list[--i];
+                            }
+                            else
+                            {
+                                list[--k] = rightElement;
+                                break;
+                            }
+                        }
+                        else
+                        {
+                            list[--k] = rightElement;
+                            if (j > 0)
+                            {
+                                rightElement = buffer[--j];
+                            }
+                            else
+                            {
+                                break;
+                            }
+                        }
+                    }
+
+                    while (j > 0)
+                    {
+                        list[--k] = buffer[--j];
+                    }
+                }  
+                position += bucketSize;
             }
-            head->length = newLength;
-            head->CheckLengthvsSize();
+            position = 0;
+            lastSize = bucketSize;
+            bucketSize *= 2;
         }
-        SetHasNoMissingValues();
-        this->InvalidateLastUsedSegment();
-        this->ClearSegmentMap();
+    }
 
-#ifdef VALIDATE_ARRAY
-        ValidateArray();
-#endif
-        return;
+    inline void SortSetHelper(Field(Var)* list, Var item, uint32 index, JsReentLock* jsReentLock, ScriptContext* scriptContext)
+    {
+        list[index] = item;
     }
 
-    uint32 JavascriptArray::sort(__inout_ecount(*len) Field(Var) *orig, uint32 *len, ScriptContext *scriptContext)
+    inline void SortSetHelper(StringItem* list, Var item, uint32 index, JsReentLock* jsReentLock, ScriptContext* scriptContext)
     {
-        uint32 count = 0, countUndefined = 0;
-        Element *elements = RecyclerNewArrayZ(scriptContext->GetRecycler(), Element, *len);
-        RecyclableObject *undefined = scriptContext->GetLibrary()->GetUndefined();
+        StringItem current;
+        current.Value = item;
+        JsReentLock lock = *jsReentLock;
+        JS_REENTRANT(lock, current.StringValue = JavascriptConversion::ToString(item, scriptContext));
+        list[index] = current;
+    }
 
-        //
-        // Create the Elements array
-        //
+    inline Var SortGetHelper(Field(Var)* list, uint32 index)
+    {
+        return list[index];
+    }
+
+    inline Var SortGetHelper(StringItem* list, uint32 index)
+    {
+        return list[index].Value;
+    }
+
+    // Helper for Array.prototype.sort
+    // template function has two paths:
+    // 1. for user supplied comparison functions (uses a list of Var)
+    // 2. for the default comparison (uses a list of StringItem)
+    template<typename T>
+    Var JavascriptArray::SortHelper(Var array, JavascriptArray::CompareVarsInfo* cvInfo)
```

Comment:
Same here, also true for related.

Reply:
Please see the extra comments I've now added here - this method is the core of my new implementation of Array.prototype.sort - it contains most of the logic to handle the changes in the spec (particularly the requirement to copy the array before sorting).